File: apps/py-hello/src/py_hello/handlers.py
from typing import Awaitable, Callable

from fastapi import APIRouter, HTTPException, Response


def health_router(readiness: Callable[[], bool]) -> APIRouter:
    r = APIRouter()

    @r.get("/healthz")
    def healthz() -> dict[str, str]:
        return {"status": "ok"}

    @r.get("/readyz")
    def readyz() -> Response:
        if not readiness():
            return Response(content='{"status":"down"}', media_type="application/json", status_code=503)
        return Response(content='{"status":"ok"}', media_type="application/json", status_code=200)

    return r


def users_router(
    create_user: Callable[[dict[str, str]], Awaitable[dict[str, str]]],
    get_user: Callable[[str], Awaitable[dict[str, str] | None]],
    list_users: Callable[[], Awaitable[list[dict[str, str]]]],
) -> APIRouter:
    r = APIRouter(prefix="/v1/users")

    @r.get("")
    async def list_endpoint() -> list[dict[str, str]]:
        return await list_users()

    @r.post("", status_code=201)
    async def create_endpoint(body: dict[str, str]) -> dict[str, str]:
        return await create_user(body)

    @r.get("/{user_id}")
    async def get_endpoint(user_id: str) -> dict[str, str]:
        u = await get_user(user_id)
        if u is None:
            raise HTTPException(status_code=404)
        return u

    return r

File: apps/rs-hello/src/handlers.rs
use std::sync::Arc;

use axum::{
    extract::Path,
    http::StatusCode,
    response::{IntoResponse, Json},
    routing::get,
    Router,
};
use serde_json::json;

pub type ReadinessFn = Arc<dyn Fn() -> bool + Send + Sync>;

pub fn health_router(ready: ReadinessFn) -> Router {
    Router::new()
        .route("/healthz", get(|| async { Json(json!({ "status": "ok" })) }))
        .route(
            "/readyz",
            get({
                let ready = ready.clone();
                move || {
                    let ready = ready.clone();
                    async move {
                        if (ready)() {
                            (StatusCode::OK, Json(json!({ "status": "ok" }))).into_response()
                        } else {
                            (StatusCode::SERVICE_UNAVAILABLE, Json(json!({ "status": "down" })))
                                .into_response()
                        }
                    }
                }
            }),
        )
}

pub fn users_router() -> Router {
    Router::new()
        .route("/v1/users", get(list_users).post(create_user))
        .route("/v1/users/:id", get(get_user))
}

async fn list_users() -> impl IntoResponse {
    Json(json!([]))
}

async fn create_user(Json(body): Json<serde_json::Value>) -> impl IntoResponse {
    (StatusCode::CREATED, Json(body))
}

async fn get_user(Path(id): Path<String>) -> impl IntoResponse {
    Json(json!({ "id": id }))
}
